CRM/Core/Permission/Joomla.php

```diff
@@ -131,4 +135,74 @@ public function checkGroupRole($array) {
     return FALSE;
   }
 
+  /**
+   * @inheritDoc
+   */
+  public function upgradePermissions($permissions) {
+    $translatedPerms = array();
+    foreach (array_flip($permissions) as $perm) {
+      $translated = $this->translateJoomlaPermission($perm);
+      $translatedPerms[] = $translated[0];
+    }
+
+    $associations = $this->getUserGroupPermsAssociations();
+    $originalAssociations = clone $associations;
+    foreach (array_keys(get_object_vars($associations)) as $permName) {
+      if (!in_array($permName, $translatedPerms)) {
+        unset($associations->$permName);
+      }
+    }
+
+    if ($originalAssociations !== $associations) {
+      $this->updateGroupPermsAssociations($associations);
+    }
+  }
+
+  /**
+   * Fetches the associations between user groups and CiviCRM permissions.
+   *
+   * @return object
+   *   Properties of the object are Joomla-fied permission names.
+   */
+  private function getUserGroupPermsAssociations() {
+    // Get a db connection.
```

Comment:
This was more about the lack of comments above.
Since you asked... comments drift, that's why. You should keep them to a minimum. I also think it is inaccurate that you do a "Reset" of the query object: there is no default query object.

I didn't know that the joomla query object could build SQL in such a way, but it was instantly understandable and I expect any mid-level developer would also recognize the pattern of: fetch dbo, set query, fetch result; it is ubiquitous.